The NEC IR protocol uses 8bit commands

Emrvb opened the pull request:
# What does this implement/fix?

The current implementation for the NEC IR protocol encoding/decoding is incomplete.

The NEC IR protocol uses 8bit ~~addresses and~~ commands. The other bits are inverses for error checking.

~~This patch will gracefully accept ~~16bit addresses and~~ commands for encoding but will emit a warning if you do. The decoder now only sets the proper 8bit ~~address and~~ command.~~

~~This patch will break existing implementations which expect to receive 16bit addresses and commands.~~

This patch will break existing implementations that rely on sending or receiving "made up" commands that fail the error checking.


## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here> WIP

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
switch:
  - platform: template
    name: Transmit NEC
    turn_on_action:
      - remote_transmitter.transmit_nec:
          address: 0x1
          command: 0xC

remote_receiver:  
  pin:
    number: ?
    inverted: ?
  dump:
    - nec
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Emrvb committed:
The NEC IR protocol uses 8bit addresses and commands

The other bits are inverses for error checking.

This patch will gracefully accept 16bit addresses and commands for encoding but will emit a warning if you do. The decoder now only sets the proper 8bit address and command.

This patch will break existing implementations which expect to receive 16bit addresses and commands.

This patch will break existing implementations that rely on sending or receiving "made up" commands that fail the error checking.

dudanov commented:
Read about the NEC Extended protocol. At least the address must remain 16-bit. And the command can be made 8-bit, but is it necessary?

Emrvb commented:
@dudanov 

You're right about the address, I guess we could leave that 16bit.

I suppose it's not necessary. It's up to the maintainers and community whether they care about this, but I for one would appreciate the validation of the command.

I guess I could also add a compatibility flag: Compatible (current behaviour), Strict-NEC, Strict-NEC-Extended.

Emrvb committed:
Fixed support for Extended NEC 16bit addresses (regression)

The Extended NEC protocol is backward compatible and the extra address space does not conflict with the original addresses. It also doesn't impose any reliability issues. For simplicity from a users' perspective it's better to transparently treat everything as if it were Extended NEC.

Emrvb committed:
Undone accidental backspace

jesserockz commented:
If we are going to break receiving by failing the checks, then we might as well break sending and only allow 8bit commands to be sent. Otherwise this has to be done eventually in the future for consistency.

Emrvb commented:
@jesserockz 

Agreed, I thought gradual deprecation might be nice for the users, but the extra work is probably not worth it.

We could also initially force 8 bit commands from the config and temporarilly allow 16bit with warning from the interface. I could even add a warning on the receiving side. But only if people deem a smoother transition necessary.

If and when this gets approved I'll update the documentation and will include this is actually Extended NEC with some remark about addressing original NEC. 

Just let me know how you want me to proceed.

jesserockz commented:
People dont always update ESPHome devices (which is not a bad thing) and so they could easily miss gradual deprecation anyway and then read only the latest notes to see why. I think its best to keep it all together.

Emrvb committed:
Removed 16bit command compatibility

Emrvb committed:
Merge branch 'esphome:dev' into nec_fix

Emrvb committed:
Fixed transmit_nec test, should use a 8bit command

Emrvb commented:
@jesserockz I totally missed your reply. Seems like dudanov made up his mind and thinks the change is necessary. I'll leave it up to you and other maintainers which patch you prefer.

I just noticed there is actually a test suite for the protocol implementations. I'll look into that for my other PRs.

ssieb commented:
There appear to be many devices that use this "protocol" with all 16 bits and the IR code references refer to them as NEC, so this will break many uses.

Emrvb commented:
@ssieb Do you have examples of those devices and their IR codes? The 16-bit codes I've come across so far are simply the 8 bit commands with their inverses tacked on. I assume this is because of the way they were dumped.

ssieb commented:
The specific codes involved were https://github.com/probonopd/irdb/blob/master/codes/Onkyo/Receiver/210%2C108.csv, but that's just an example.  There's also more discussion at https://github.com/esphome/issues/issues/3309.

Emrvb commented:
@ssieb Your Onkyo example is Extended NEC, using 16-bit addressing and 8-bit commands. Perhaps I can help you making those work if you can describe where you're stuck (open a new issue for that and mention this merge so I can find it). 

In the other issue you linked it seems to me there is some confusion about what the dumped codes mean, but I'm pretty confident that's just regular Extended NEC. 

There are some variants that defer slightly (probably compatible) and ways of handling repeats (currently not supported). What I also found is that for example Arduino-IRremote originally accepted a raw payload so that you could send and receive random data (you can't do that anymore, they also use 16-bit addressing and 8-bit commands now). That makes it not NEC and we should protect users from doing that. For that use case we should decide on a data protocol with some error checking to implement.

Emrvb commented:
@ssieb I'm sorry, I think we're not quite understanding each other. Based on the links you shared I don't see a problem. Would you mind sharing an example config for that Onkyo Receiver that you reckon would break by this change?

ssieb commented:
```yaml
  - remote_transmitter.transmit_nec:
      address: 0x6dd2
      command: 0xfb04
```

Emrvb commented:
@ssieb Your example shows me you're using an Onkyo (device 0xD2 / 210 and subdevice 0x6D / 109). If I'm not mistaken that is a slightly different specification called NEC1. It basically mirrors Extended NEC but the addressing rules are slightly different.

The command you issue is 0x04 / 4, Power. The 0xFB (251) is the inverse of that command for error checking and not actually part of the command.

0x04 => 00000100
0xFB => 11111011

The CSV you linked specifies to use the command 4, not 64260. This patch allows you to specify the correct command, 4, and it wil automatically add the correct inverse. 

Like I said before:
>  The 16-bit codes I've come across so far are simply the 8 bit commands with their inverses tacked on. I assume this is because of the way they were dumped.

This patch will break existing implementations, as is mentioned in the conversation above, by no longer accepting the redundant inverses with the command. Users will have to fix their configuration/implementations by removing the "erroneous" byte, after which everything should work like before.

I hope I've clarified this enough to resolve your objections to this patch. Let me know when that's not the case.

ssieb commented:
Ok, I read through the discussion again and I see I missed the part about allowing 16-bit addresses.  Maybe edit the initial comment to include that information.  If that's included, then I don't have any objections.

Emrvb commented:
I still think this should be merged.
